app/src/org/commcare/preferences/AdvancedActionsPreferences.java

```diff
@@ -86,6 +107,26 @@ public class AdvancedActionsPreferences extends CommCarePreferenceFragment {
         keyToTitleMap.put(ENABLE_MANUAL_FORM_QUARANTINE, "menu.enable.manual.form.quarantine");
     }
 
+    public enum PermissionAction{
+        NO_ACTION,
+        REJECT_FEATURE,
+        REQUEST_PERMISSION;
+    }
+
+    @Override
+    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
+        super.onCreatePreferences(savedInstanceState, rootKey);
+
+        showingPermissionDialog.observe(this, value -> {
+            if (value == REJECT_FEATURE){
+                informUserAboutFeatureUnavailability();
+            }
+            else if (value == REQUEST_PERMISSION){
+                showNearbyWiFiPermissionRationale();
+            }
+        });
```

Comment:
@shubham1g5 I battled with this a bit, the reasoning was to prevent the creation of `CommCareWiFiDirectActivity` ini case the permission was not granted, which happens in the `AdvancedActionsPreference`. It does feel misplaced having it there, but within `CommCareWiFiDirectActivity` we will need to put behind the `Transfer` and `Receive` options.

Reply:
Can't we simply add the permission check in `onCreate` for  `CommCareWiFiDirectActivity`  and show user an error if the permission is not granted on the same screen ?

Reply:
@shubham1g5 we can, but where should we trigger the permission request workflow, in case the user hasn't yet been asked?

Reply:
[55b5545](https://github.com/dimagi/commcare-android/pull/2683/commits/55b5545d9fba3b18f5e394c3a4c1168f3d9e8d94)